### PR TITLE
Add Changelog component

### PR DESCRIPTION
**Pull request by myarmolinsky**
Add Changelog component

Resolves: #1623 
Change-type: minor
Signed-off-by: Matthew Yarmolinsky <matthew-timothy@balena.io>

---

##### Contributor checklist

- [ ] I have regenerated jest snapshots for any affected components with `npm run generate-snapshots`

##### Reviewer Guidelines

- When submitting a review, please pick:
  - '_Approve_' if this change would be acceptable in the codebase (even if there are minor or cosmetic tweaks that could be improved).
  - '_Request Changes_' if this change would not be acceptable in our codebase (e.g. bugs, changes that will make development harder in future, security/performance issues, etc).
  - '_Comment_' if you don't feel you have enough information to decide either way (e.g. if you have major questions, or you don't understand the context of the change sufficiently to fully review yourself, but want to make a comment)

---

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -9,6 +9,8 @@ export {
 	ButtonGroupOption,
 } from './components/ButtonGroup';
 export { Card, CardProps } from './components/Card';
+export { Changelog, ChangelogProps } from './components/Changelog';
+export { default as processChangelog } from './extra/Changelog/helpers.js';
```
Comment (thgreasi) [inline src/index.ts:13]: The point of having it in the `extra`s is to not export it via the `index.ts`.
Let's either remove it from here, or move it outside the `extra`s.

File: src/extra/Changelog/helpers.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+export default (data) => {
+	const untilDate = new Date(new Date().setDate(new Date().getDate() - 90));
+	const dateStringsIterator =
+		data.matchAll(
+			/\n# v\d+\.\d+\.\d+\n## \(([0-9]{4}-[0-1][0-9]-[0-3][0-9])\)/g
```
Comment (Page-) [inline src/extra/Changelog/helpers.js:5]: You could use multiline mode + `^` instead:
```suggestion
			/^# v\d+\.\d+\.\d+\n## \(([0-9]{4}-[0-1][0-9]-[0-3][0-9])\)/gm
```
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#advanced_searching_with_flags

File: src/components/Changelog/ChangelogButton.tsx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+export const ChangelogButton = ({
+	getChangelog,
+	onClick,
+	latestChangelogVersion,
+	style,
+	...restProps
+}: ChangelogProps &
+	ButtonProps & {
+		latestChangelogVersion?: string;
+		onClick?: () => void;
```
Comment (JSReds) [inline src/components/Changelog/ChangelogButton.tsx:32]: onClick prop is already part of `ButtonProps`, no need to duplicate it

File: src/components/Changelog/ChangelogButton.tsx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+		[showChangelogModal],
+	);
+
+	return (
+		<Flex justifyContent="flex-end">
+			<Button
+				height="38px"
+				color="quartenary.main"
+				onClick={() => {
+					onClick?.();
```
Comment (JSReds) [inline src/components/Changelog/ChangelogButton.tsx:49]:
```suggestion
				onClick={(e) => {
					onClick?.(e);
```

File: src/components/Changelog/ChangelogButton.tsx [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+			getFromLocalStorage<string | undefined>(
+				LAST_VIEWED_CHANGELOG_VERSION_STORAGE_KEY,
+			),
+		[showChangelogModal],
+	);
+
+	return (
+		<Flex justifyContent="flex-end">
+			<Button
+				height="38px"
```
Comment (JSReds) [inline src/components/Changelog/ChangelogButton.tsx:46]: let's use https://github.com/balena-io-modules/rendition/blob/ff4c9e135b7bd4936b5793be5e22eef058f65161/src/theme.ts#L114

File: src/components/Changelog/index.tsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+const markdownComponentOverrides: React.ComponentProps<
+	typeof Markdown
+>['componentOverrides'] = {
+	a: (props) => <Link blank {...props} />,
+	h1: (props) => <Heading.h1 {...props} fontSize="20px" />,
+	h2: (props) => <Heading.h2 {...props} fontSize="16px" />,
+	h3: (props) => <Heading.h3 {...props} fontSize="14px" />,
+	h4: (props) => <Heading.h4 {...props} fontSize="12px" />,
```
Comment (JSReds) [inline src/components/Changelog/index.tsx:19]: let's try to use the theme when possible 
https://github.com/balena-io-modules/rendition/blob/ff4c9e135b7bd4936b5793be5e22eef058f65161/src/theme.ts#L119

File: src/index.ts [hunk 7]
```diff
@@ -9,6 +9,7 @@ export {
 	ButtonGroupOption,
 } from './components/ButtonGroup';
 export { Card, CardProps } from './components/Card';
+export { Changelog, ChangelogProps } from './components/Changelog';
```
Comment (JSReds) [inline src/index.ts:12]: let's export the `processChangelog` function as well so we don't have to do any deep import

Comment (myarmolinsky) [reply src/index.ts:12]: Resolved here: https://balena.zulipchat.com/#narrow/stream/346007-balena-io.2FbalenaCloud/topic/Changelog.20-.3E.20rendition/near/308960660

File: tsconfig.esm5.json [hunk 8]
```diff
@@ -1,7 +1,7 @@
 {
   "extends": "./tsconfig.json",
   "compilerOptions": {
-    "module": "es2015",
+    "module": "es2020",
```
[the remark this answers is not in the document]
Comment (thgreasi) [reply tsconfig.esm5.json:4]: This is a possibly breaking change b/c it changes the following in index.js:
```
import * as SchemaSieve_1 from './components/Filters/SchemaSieve';
export { SchemaSieve_1 as SchemaSieve };
```
which now becomes
```
export * as SchemaSieve from './components/Filters/SchemaSieve';
```
I don't really expect it to cause issues to anyone though, but if you want to be super safe you could replace the `export *` we have in the src with the above import & export combination and leave a note to change it in the next major.

**Comment by myarmolinsky** [comment]
@balena-ci I self certify!

**Pull request merged**